cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -791,6 +854,22 @@ def configure_master_services():
     api_opts.add('insecure-bind-address', '127.0.0.1')
     api_opts.add('insecure-port', '8080')
     api_opts.add('storage-backend', 'etcd2')  # FIXME: add etcd3 support
+
+    # TODO: investigate if it's possible to use config file to store args
+    # Handle api-extra-args config option
+    to_add, to_remove = get_config_args()
+    for arg in to_add:
+        hookenv.log('Adding option: {} {}'.format(arg[0], arg[1]))
+        # Make sure old value is gone
+        api_opts.destroy(arg[0])
+        api_opts.add(arg[0], arg[1])
+    for arg in to_remove:
+        hookenv.log('Removing option: {}'.format(arg))
+        api_opts.destroy(arg)
+        # We need to "unset" options by settig their value to "null" string
+        cmd = ['snap', 'set', 'kube-apiserver', '{}=null'.format(arg)]
+        check_call(cmd)
```

Comment:
> My feeling is that we should make it {'foo': ['baz']} and trust that the operator knows what they are doing.

:+1: to this

> An example might be "--min-request-timeout" which the charm sets to 300s. The operator might set it to 600s, realize it's too short and "unset" to revert expecting the flag to be removed (and in consequence having k8s default value of 1800s).

Intuition tells me that `juju config kubernetes-master api-extra-args=""` should give the same result no matter what it was set to prior. I think that behavior is important.

If you take that stance, then I think it becomes clear: removing a value from api-extra-args _has_ to revert to charm default (300s in your example). If the operator really wants to bypass the charm default, they can do it by overriding with an explicit value.

That said, I don't see a quick fix for this with how the charm uses FlagManager today. Seems like there might be a lot of refactoring involved to get it right. :\